src/datasets/arrow_dataset.py

```diff
@@ -2460,6 +2460,18 @@ def to_pandas(
                 for offset in range(0, len(self), batch_size)
             )
 
+    def add_column(self, column: dict):
+        """Add column to Dataset.
+
+        Args:
+            column (dict): Column data to be added.
+        """
+        column_table = InMemoryTable.from_pydict(column)
+        # Concatenate tables horizontally
+        self._data = ConcatenationTable.from_blocks([[self._data, column_table]])
```

Comment:
`self._data` may not be a valid table block (i.e. either a InMemoryTable or a MemoryMappedTable object).
For example if `self._data` is a `ConcatenationTable`, this won't work.

Maybe we can use another `ConcatenationTable` constructor for this ?
Like for example a version of `ConcatenationTable.from_tables` but for axis=1 ?
Under the hood this uses `from_blocks` anyway, but it allows any kind of tables as input.

Reply:
I am going to finish this PR first: #2151 😉